### PR TITLE
Infer the array size for slice patterns

### LINKED ISSUE #11478: Wrong variable type was hinted
**rust-analyzer version**: 9bc2ee34b 2022-02-15 dev

**rustc version**: rustc 1.60.0-nightly (1e12aef3f 2022-02-13)

Code:

```rust
let [head, tail @ ..] =  [1, 2, 3, 4];
```

RA hints:
```rust
let [head: i32, tail: [i23; 4] @ ..] =  [1, 2, 3, 4];
```

Expected:
```rust
let [head: i32, tail: [i23; 3] @ ..] =  [1, 2, 3, 4];
```

### PR DESCRIPTION
**Pull request by ChayimFriedman2**
Fixes #11478 (feat or fix?),

I don't know much about the type-checking in RA, so maybe I did some obvious mistake.

### CODE REVIEW

File: crates/hir_ty/src/infer/pat.rs [hunk 1]
```diff
@@ -232,16 +233,33 @@ impl<'a> InferenceContext<'a> {
...
+                                ConstScalar::Usize(length) => {
+                                    let length =
+                                        length.checked_sub((prefix.len() + suffix.len()) as u64);
+                                    TyKind::Array(
+                                        elem_ty.clone(),
+                                        crate::consteval::usize_const(length),
+                                    )
+                                    .intern(Interner)
+                                }
+                                ConstScalar::Unknown => self.err_ty(),
```
Comment (flodiebold) [inline crates/hir_ty/src/infer/pat.rs:249]: this should not result in err_ty, rather it should just reuse the length value as before.

File: crates/hir_ty/src/infer/pat.rs [hunk 2]
```diff
@@ -232,16 +233,33 @@ impl<'a> InferenceContext<'a> {
...
+                            },
+                            _ => self.err_ty(),
```
Comment (flodiebold) [inline crates/hir_ty/src/infer/pat.rs:251]: here as well

File: crates/hir_ty/src/infer/pat.rs (hunk 1 above)
Comment (ChayimFriedman2) [reply crates/hir_ty/src/infer/pat.rs]: Made them return an unresolved constant (but array of the correct type), because it's a type error.

**Review by flodiebold — COMMENTED** [review]
bors r+

**Pull request merged**